src/main/java/edu/harvard/dbmi/avillach/dictionaryetl/facet/FacetLoaderService.java

```diff
@@ -0,0 +1,429 @@
+
+            // Insert union of direct children’s concept mappings into the parent
+            mapped = facetConceptRepository.mapParentToUnionOfDirectChildren(facet.getFacetId());
+        }
+
+        accum.facetMappings().add(new FacetMappingBreakdown(categoryName, name, mapped));
+        return new Counts(created, updated);
+    }
+
+    /** Legacy mapping by flat AND list. Implementation assumed to exist previously. */
+    private long mapFacetToConcepts(Long facetId, List<FacetExpressionDTO> effective) {
+        // Existing implementation assumed; if not present, implement a simple version here.
+        // For safety, implement a streaming mapper identical in semantics to legacy evaluator.
+
+        facetConceptRepository.deleteAllForFacetIds(java.util.List.of(facetId)); // ensure clean state before remap
+
+        long inserted = 0;
+        try (java.util.stream.Stream<ConceptPathRow> rows = conceptRepository.streamLeafNodeIdAndPath()) {
+            List<Long> batch = new ArrayList<>(1024);
+            for (Iterator<ConceptPathRow> it = rows.iterator(); it.hasNext(); ) {
+                ConceptPathRow row = it.next();
+                if (FacetExpressionEvaluator.facetAppliesToConceptPath(effective, row.getConceptPath())) {
+                    batch.add(row.getConceptNodeId());
+                    if (batch.size() >= 1000) {
+                        inserted += facetConceptRepository.bulkMap(facetId, batch);
+                        batch.clear();
+                    }
+                }
+            }
+            if (!batch.isEmpty()) {
+                inserted += facetConceptRepository.bulkMap(facetId, batch);
+            }
+        }
+        return inserted;
+    }
+
+    /** New mapping for OR-of-AND groups. */
+    private long mapFacetToConceptsGrouped(Long facetId, List<List<FacetExpressionDTO>> groups) {
+        // Strategy:
+        // - Clear existing
+        // - For each concept, accept if ANY group matches (OR)
+        // - Bulk insert mappings
+        facetConceptRepository.deleteAllForFacetIds(java.util.List.of(facetId)); // ensure clean state before remap
+
+        long inserted = 0;
+        try (java.util.stream.Stream<ConceptPathRow> rows = conceptRepository.streamLeafNodeIdAndPath()) {
+            List<Long> batch = new ArrayList<>(1024);
+            for (Iterator<ConceptPathRow> it = rows.iterator(); it.hasNext(); ) {
+                ConceptPathRow row = it.next();
+                if (FacetExpressionEvaluator.facetAppliesToConceptPathGrouped(groups, row.getConceptPath())) {
+                    batch.add(row.getConceptNodeId());
+                    if (batch.size() >= 1000) {
+                        inserted += facetConceptRepository.bulkMap(facetId, batch);
+                        batch.clear();
+                    }
+                }
+            }
+            if (!batch.isEmpty()) {
+                inserted += facetConceptRepository.bulkMap(facetId, batch);
```

Comment:
Can we bump to Java 24 for gatherers? It makes this kind of batching stuff a lot simpler.
```suggestion
               facetConceptRepository.deleteAllForFacetIds(java.util.List.of(facetId)); // ensure clean state before remap
        return conceptRepository.streamLeafNodeIdAndPath()
            .filter(row -> FacetExpressionEvaluator.facetAppliesToConceptPath(effective, row.getConceptPath()))
            .map(ConceptPathRow::getConceptNodeId)
            .gather(Gatherers.windowFixed(1024))
            .mapToInt(batch -> facetConceptRepository.bulkMap(facetId, batch))
            .sum();
```